dask/array/blockwise.py

```diff
@@ -3,10 +3,106 @@
 
 import tlz as toolz
 
+from collections.abc import Mapping
+from functools import reduce
+from itertools import product
+
 from .. import base, utils
 from ..delayed import unpack_collections
 from ..highlevelgraph import HighLevelGraph
-from ..blockwise import blockwise as core_blockwise
+from ..blockwise import BlockwiseIO, blockwise as core_blockwise
+
+
+class CreateArraySubgraph(Mapping):
+    """
+    Subgraph for numpy array creation utilities
+    """
+
+    def __init__(
+        self,
+        name,
+        func,
+        shape,
+        chunks,
+    ):
+        self.name = name
+        self.func = func
+        self.shape = shape
+        self.chunks = chunks
+        # Compute total number of blocks
+        self._len = reduce(lambda acc, x: len(x) * acc, chunks, 1)
+
+    def __getitem__(self, key):
+        """
+        Construct a task for a given key.
+        """
+        try:
+            name, *idx = key
+        except ValueError:
+            # too many / few values to unpack
+            raise KeyError(key) from None
+
+        if name != self.name:
+            raise KeyError(key)
+
+        # Check for invalid indices
+        if any(i < 0 or i >= len(c) for i, c in zip(idx, self.chunks)):
+            raise KeyError(key)
+
+        # Get this chunk shape
+        ndim = len(self.chunks)
+        chunk = tuple(self.chunks[i][idx[i]] for i in range(ndim))
+
+        return (self.func, chunk)
```

Comment:
This looks correct, and makes sense - You are using `partial` to wrap all options into `self.func`, because it is the same for every chunk.  The only argument that may vary between two different chunks is the shape of that "chunk".  Please correct me if I am missing anything :)

My suggestion is to add a sentence or two in the docstring to explain/document this assumption.

Reply:
@rjzamora You are right -- I'm currently thinking through whether this needs to be more flexible for other functions than initially implemented here. At the end of the day, we need to be able to ascertain everything we need from the `key`, but more info may need to be bound to any given `BlockwiseIO` to get stuff like `from_array` working.

Reply:
> but more info may need to be bound to any given BlockwiseIO to get stuff like from_array working.

Good point- `CreateArraySubgraph` + `BlockwiseCreateArray` does not target the case where you have input data.  I'm unsure if we want to use the same code path for `from_array` (maybe we do).  For now, I'm okay with the solution focusing on pure "creation,"

Reply:
I'm wondering if it would make sense to pull out something like this logic for [`block_info`](https://github.com/dask/dask/blob/1d669480f78efa290fad612d196ecd18d26e7972/dask/array/core.py#L735-L809) and make it an optional argument to `func`.